# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入数据分析所需要的库，并通过Pandas的`read_csv`函数，将原始数据文件"e_commerce.csv"里的数据内容，解析为DataFrame，并赋值给变量`original_data`。

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

在这一部分，我将对在上一部分建立的`original_data`这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

### 评估数据整齐度

In [4]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
10477     537237     22271                FELTCRAFT DOLL ROSIE         1   
411839    572234     23534               STOP FOR TEA WALL ART         3   
402532    571502     21922   UNION STRIPE WITH FRINGE  HAMMOCK         4   
165910    550835     22228     BUNNY WOODEN PAINTED WITH BIRD          4   
393321    570818     22720   SET OF 3 CAKE TINS PANTRY DESIGN          2   
370282    569134    85159B         WHITE TEA,COFFEE,SUGAR JARS        12   
367792    568908     22077              6 RIBBONS RUSTIC CHARM        12   
129649    547386     84988  SET OF 72 PINK HEART PAPER DOILIES         2   
277955    561201     23080            RED METAL BOX TOP SECRET         2   
198521    554054     22082         RIBBON REEL STRIPES DESIGN          1   

             InvoiceDate  UnitPrice  CustomerID         Country  
10477     12/6/2010 9:58       5.91         NaN  United Kingdom  
411839  10/21/2011 14:41       5.95     16293.0  United Kingdom  
402532  10/17/2011 15:18       7.95     14422.0  United Kingdom  
165910   4/21/2011 10:52       0.85     15034.0  United Kingdom  
393321  10/12/2011 12:47       4.95     17841.0  United Kingdom  
370282   9/30/2011 14:07       1.95     15411.0  United Kingdom  
367792   9/29/2011 13:38       1.95     16653.0  United Kingdom  
129649   3/22/2011 15:52       2.92         NaN  United Kingdom  
277955   7/25/2011 15:08       8.25     14692.0  United Kingdom  
198521   5/20/2011 15:29       3.29         NaN  United Kingdom

从抽样的10行数据数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

### 评估数据干净度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
InvoiceNo      541909 non-null object
StockCode      541909 non-null object
Description    540455 non-null object
Quantity       541909 non-null int64
InvoiceDate    541909 non-null object
UnitPrice      541909 non-null float64
CustomerID     406829 non-null float64
Country        541909 non-null object
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，而`Description`、`CustomerID`变量存在缺失值。

此外，`InvoiceDate`的数据类型应为日期，`CustomerID`的数据类型应为字符串，应当进行数据格式转换。

#### 评估缺失数据

在了解`Description`存在缺失值后，根据条件提取出缺失观察值。

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity       InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56   12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1   12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1   12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1   12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1   12/1/2010 14:34        0.0   
1988      536550     85044         NaN         1   12/1/2010 14:34        0.0   
2024      536552     20950         NaN         1   12/1/2010 14:34        0.0   
2025      536553     37461         NaN         3   12/1/2010 14:35        0.0   
2026      536554     84670         NaN        23   12/1/2010 14:35        0.0   
2406      536589     21777         NaN       -10   12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38   12/2/2010 14:42        0.0   
4348      536765    84952C         NaN        19   12/2/2010 14:43        0.0   
7187      536995     35951         NaN        57   12/3/2010 15:29        0.0   
7188      536996     22712         NaN       -20   12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20   12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6   12/3/2010 15:30        0.0   
7191      536999     21421         NaN       110   12/3/2010 15:32        0.0   
7192      537000     21414         NaN       -22   12/3/2010 15:32        0.0   
7193      537001     21653         NaN        -6   12/3/2010 15:33        0.0   
7194      537002     21651         NaN        70   12/3/2010 15:33        0.0   
7195      537003     85126         NaN        -2   12/3/2010 15:33        0.0   
7196      537004     21814         NaN       -30   12/3/2010 15:34        0.0   
7197      537005     21692         NaN       -70   12/3/2010 15:35        0.0   
7198      537006     21648         NaN      -130   12/3/2010 15:36        0.0   
7199      537007     21172         NaN       -80   12/3/2010 15:36        0.0   
7200      537008     21161         NaN      -120   12/3/2010 15:37        0.0   
7201      537009    84534B         NaN       -80   12/3/2010 15:38        0.0   
7202      537010     22162         NaN       -40   12/3/2010 15:38        0.0   
7203      537011     22139         NaN        -5   12/3/2010 15:38        0.0   
7204      537012     21535         NaN       -12   12/3/2010 15:39        0.0   
...          ...       ...         ...       ...               ...        ...   
497632    578431     22579         NaN       -18  11/24/2011 12:15        0.0   
497677    578456     22579         NaN        11  11/24/2011 12:29        0.0   
498597    578529     22087         NaN       -23  11/24/2011 14:29        0.0   
499156    578628     23541         NaN        20  11/24/2011 15:55        0.0   
499157    578629     22943         NaN        24  11/24/2011 15:56        0.0   
499169    578632     22579         NaN         4  11/24/2011 16:22        0.0   
499787    578693     21811         NaN         1   11/25/2011 9:38        0.0   
500161    578738     21638         NaN       -33  11/25/2011 11:33        0.0   
508796    579272     22812         NaN         2  11/29/2011 10:00        0.0   
510622    579460     21110         NaN      -151  11/29/2011 14:39        0.0   
511451    579500     22810         NaN        33  11/29/2011 16:14        0.0   
514205    579688     21116         NaN       -33  11/30/2011 13:51        0.0   
514287    579691    47503A         NaN        -5  11/30/2011 14:05        0.0   
514651    579738     21927         NaN         8  11/30/2011 14:26        0.0   
516189    579880     22947         NaN         5  11/30/2011 17:14        0.0   
517719    580087    84341B         NaN        -2   12/1/2011 14:27        0.0   
519967    580197     22696         NaN         6   12/2/2011 12:04        0.0   
521333    580359     20775         NaN

有1454条交易数据缺失`Description`变量值。

从输出结果来看，这些缺失`Description`的交易数据，`UnitPrice`都为0。为了验证猜想，我们增加筛选条件，看是否存在`Description`变量缺失且`UnitPrice`不为0的数据。

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来结果数量为0条，说明缺失`Description`值的数据，同时也不具备有效的`UnitPrice`值。

`Description`表示产品名称，`UnitPrice`表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

`CustomerID`变量同样存在缺失值，因此也根据条件提取出缺失观察值。

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                          Description  Quantity  \
622       536414     22139                                  NaN        56   
1443      536544     21773      DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774      DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786                   POLKADOT RAIN HAT          4   
1446      536544     21787                RAIN PONCHO RETROSPOT         2   
1447      536544     21790                   VINTAGE SNAP CARDS         9   
1448      536544     21791   VINTAGE HEADS AND TAILS CARD GAME          2   
1449      536544     21801  CHRISTMAS TREE DECORATION WITH BELL        10   
1450      536544     21802      CHRISTMAS TREE HEART DECORATION         9   
1451      536544     21803       CHRISTMAS TREE STAR DECORATION        11   
1452      536544     21809     CHRISTMAS HANGING TREE WITH BELL         1   
1453      536544     21810     CHRISTMAS HANGING STAR WITH BELL         3   
1454      536544     21811    CHRISTMAS HANGING HEART WITH BELL         1   
1455      536544     21821     GLITTER STAR GARLAND WITH BELLS          1   
1456      536544     21822    GLITTER CHRISTMAS TREE WITH BELLS         1   
1457      536544     21823  PAINTED METAL HEART WITH HOLLY BELL         2   
1458      536544     21844                    RED RETROSPOT MUG         2   
1459      536544     21851       LILAC DIAMANTE PEN IN GIFT BOX         1   
1460      536544     21870     I CAN ONLY PLEASE ONE PERSON MUG         1   
1461      536544     21871                  SAVE THE PLANET MUG         5   
1462      536544     21874                    GIN AND TONIC MUG         1   
1463      536544     21879                     HEARTS GIFT TAPE         1   
1464      536544     21884            CAKES AND BOWS GIFT  TAPE         1   
1465      536544     21888                            BINGO SET         1   
1466      536544     21889               WOODEN BOX OF DOMINOES         2   
1467      536544     21892   TRADITIONAL WOODEN CATCH CUP GAME          3   
1468      536544     21894          POTTING SHED SEED ENVELOPES         1   
1469      536544     21911                   GARDEN METAL SIGN          1   
1470      536544     21912             VINTAGE SNAKES & LADDERS         3   
1471      536544     21913       VINTAGE SEASIDE JIGSAW PUZZLES         1   
...          ...       ...                                  ...       ...   
541511    581498     71053         WHITE MOROCCAN METAL LANTERN         1   
541512    581498    72349b      SET/6 PURPLE BUTTERFLY T-LIGHTS         2   
541513    581498     79321                        CHILLI LIGHTS        10   
541514    581498    82001s            SILVER RECORD COVER FRAME         2   
541515    581498     82482    WOODEN PICTURE FRAME WHITE FINISH         4   
541516    581498     82552                  WASHROOM METAL SIGN         1   
541517    581498     82580                  BATHROOM METAL SIGN         1   
541518    581498     82581                    TOILET METAL SIGN         1   
541519    581498     82600                N0 SINGING METAL SIGN         4   
541520    581498    84029E       RED WOOLLY HOTTIE WHITE HEART.         4   
541521    581498    84032A   CHARLIE+LOLA PINK HOT WATER BOTTLE         4   
541522    581498    84032B  CHARLIE + LOLA RED HOT WATER BOTTLE         3   
541523    581498     84375        SET OF 20 KIDS COOKIE CUTTERS         3   
541524    581498    84509a      SET OF 4 ENGLISH ROSE PLACEMATS         1   
541525    581498    84558a         3D DOG PICTURE PLAYING CARDS         1   
541526    581498     84832     ZINC WILLIE WINKIE  CANDLE STICK        26   
541527    581498    84968e      SET OF 16 VINTAGE BLACK CUTLERY         1   
541528    581498    84970s    HANGING HEART ZINC T-LIGHT HOLDER         1   
541529    581498    84997a    CHILDRENS CUTLERY POLKADOT GREEN          2   
541530    581498    84997b     CHILDRENS CUTLERY RETROSPOT RED          3   
541

`CustomerID`表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些`CustomerID`缺失的交易数据仍然有效，因此保留此变量为空的观察值。

#### 评估重复数据

根据数据变量的含义来看，虽然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一标识符，但一次交易可能包含多件商品，因此`InvoiceNo`可以存在重复，不同交易可以包含同件商品，因此`StockCode`可以存在重复，顾客可以进行多次交易或下单多个商品，因此`CustomerID`也可以存在重复。

那么针对此数据集，我们无需评估重复数据。

#### 评估不一致数据

不一致数据可能存在于`Country`变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [9]:
original_data["Country"].value_counts()

United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68
European

从`Country`变量值来看，`"USA"`、`"United States"`均在表示美国，`"United Kingdom"`、`"UK"`、`"U.K."`均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

#### 评估无效或错误数据

可以通过DataFrame的`describe`方法，对数值统计信息进行快速了解。

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，`Quantity`和`UnitPrice`存在负数，会对后续数值分析产生影响。

因此，我们先筛选出`Quantity`数值为负数的观察值，进一步评估其含义。

In [11]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                          Description  Quantity  \
141      C536379         D                             Discount        -1   
154      C536383    35004C      SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556       PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984     PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983     PACK OF 12 BLUE PAISLEY TISSUES        -24   
238      C536391     21980    PACK OF 12 RED RETROSPOT TISSUES        -24   
239      C536391     21484          CHICK GREY HOT WATER BOTTLE       -12   
240      C536391     22557     PLASTERS IN TIN VINTAGE PAISLEY        -12   
241      C536391     22553               PLASTERS IN TIN SKULLS       -24   
939      C536506     22960             JAM MAKING SET WITH JARS        -6   
1441     C536543     22632            HAND WARMER RED RETROSPOT        -1   
1442     C536543     22355            CHARLOTTE BAG SUKI DESIGN        -2   
1973     C536548     22244           3 HOOK HANGER MAGIC GARDEN        -4   
1974     C536548     22242        5 HOOK HANGER MAGIC TOADSTOOL        -5   
1975     C536548     20914  SET/5 RED RETROSPOT LID GLASS BOWLS        -1   
1976     C536548     22892    SET OF SALT AND PEPPER TOADSTOOLS        -7   
1977     C536548     22654                   DELUXE SEWING KIT         -1   
1978     C536548     22767          TRIPLE PHOTO FRAME CORNICE         -2   
1979     C536548     22333    RETROSPOT PARTY BAG + STICKER SET        -1   
1980     C536548     22245         HOOK, 1 HANGER ,MAGIC GARDEN        -2   
1981     C536548     22077               6 RIBBONS RUSTIC CHARM        -6   
1982     C536548     22631             CIRCUS PARADE LUNCH BOX         -1   
1983     C536548     22168        ORGANISER WOOD ANTIQUE WHITE         -2   
1984     C536548     21218               RED SPOTTY BISCUIT TIN        -3   
1985     C536548     20957         PORCELAIN HANGING BELL SMALL        -1   
1986     C536548     22580         ADVENT CALENDAR GINGHAM SACK        -4   
2406      536589     21777                                  NaN       -10   
3170     C536606     20914  SET/5 RED RETROSPOT LID GLASS BOWLS        -2   
3338     C536622     22752         SET 7 BABUSHKA NESTING BOXES        -2   
3410     C536625     22839      3 TIER CAKE TIN GREEN AND CREAM        -2   
...          ...       ...                                  ...       ...   
538565   C581409     22173   METAL 4 HOOK HANGER FRENCH CHATEAU        -2   
538566   C581409    85199L  LARGE HANGING IVORY & RED WOOD BIRD        -1   
538567   C581409     85127   SMALL SQUARE CUT GLASS CANDLESTICK        -5   
538919    581422     23169                              smashed      -235   
540072   C581460     22197                       POPCORN HOLDER        -5   
540073   C581460     22107                   PIZZA PLATE IN BOX        -1   
540078   C581462     16219         HOUSE SHAPE PENCIL SHARPENER       -48   
540079   C581462     21642            ASSORTED TUTTI FRUTTI PEN       -72   
540080   C581463     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        -2   
540081   C581464     23458             DOLLY CABINET 3 DRAWERS         -1   
540082   C581464     71477   COLOURED GLASS STAR T-LIGHT HOLDER        -6   
540083   C581465     23660                   HENRIETTA HEN MUG         -2   
540084   C581465     22171     3 HOOK PHOTO SHELF ANTIQUE WHITE        -1   
540085   C581465     21876                        POTTERING MUG        -4   
540086   C581465     20914  SET/5 RED RETROSPOT LID GLASS BOWLS        -3   
540087   C581466     22838        3 TIER CAKE TIN RED AND CREAM        -1   
540088   C581466     22720    SET OF 3 CAKE TINS PANTRY DESIGN         -2   
540089   C581466     21216     SET 3 RETROSPOT TEA,COFFEE,SUGAR        -1   
540090   C581466     21535         RED RETROSPOT SMALL MILK JUG        -2   
540091   C581466     21232       STRAWBERRY CERAMIC TRINKET POT        -1   
540

从筛选结果来看，`Quantity`变量为负数的观察值，`InvoiceNo`似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

In [12]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
2406      536589     21777              NaN       -10  12/1/2010 16:50   
4347      536764    84952C              NaN       -38  12/2/2010 14:42   
7188      536996     22712              NaN       -20  12/3/2010 15:30   
7189      536997     22028              NaN       -20  12/3/2010 15:30   
7190      536998     85067              NaN        -6  12/3/2010 15:30   
7192      537000     21414              NaN       -22  12/3/2010 15:32   
7193      537001     21653              NaN        -6  12/3/2010 15:33   
7195      537003     85126              NaN        -2  12/3/2010 15:33   
7196      537004     21814              NaN       -30  12/3/2010 15:34   
7197      537005     21692              NaN       -70  12/3/2010 15:35   
7198      537006     21648              NaN      -130  12/3/2010 15:36   
7199      537007     21172              NaN       -80  12/3/2010 15:36   
7200      537008     21161              NaN      -120  12/3/2010 15:37   
7201      537009    84534B              NaN       -80  12/3/2010 15:38   
7202      537010     22162              NaN       -40  12/3/2010 15:38   
7203      537011     22139              NaN        -5  12/3/2010 15:38   
7204      537012     21535              NaN       -12  12/3/2010 15:39   
7205      537013     35965              NaN       -25  12/3/2010 15:40   
7206      537014     21040              NaN       -20  12/3/2010 15:40   
7207      537015     37464              NaN       -14  12/3/2010 15:41   
7208      537016     21836              NaN       -15  12/3/2010 15:41   
7209      537017    47556B              NaN       -69  12/3/2010 15:41   
7210      537018    84562A              NaN        -3  12/3/2010 15:42   
7211      537019     37474              NaN       -15  12/3/2010 15:43   
7212      537020     21807              NaN       -30  12/3/2010 15:43   
7213      537021     21011              NaN       -20  12/3/2010 15:43   
7291      537027    18098C              NaN      -140  12/3/2010 16:36   
7292      537028     21696              NaN      -320  12/3/2010 16:36   
7293      537029     37476              NaN        -6  12/3/2010 16:37   
7313      537032     21275                ?       -30  12/3/2010 16:50   
...          ...       ...              ...       ...              ...   
517719    580087    84341B              NaN        -2  12/1/2011 14:27   
519986    580218     22577            check      -300  12/2/2011 12:12   
521333    580359     20775              NaN       -35  12/2/2011 16:11   
522158    580377     84609          damages       -29  12/2/2011 17:52   
522159    580378     22792          damages       -73  12/2/2011 17:56   
522160    580379    72225C              NaN      -144  12/2/2011 17:57   
522161    580380     21620            check       -20  12/2/2011 17:58   
522162    580381     21758              NaN        -9  12/2/2011 17:58   
522163    580382     16045            check      -900  12/2/2011 17:59   
522164    580383     20652            check       -37  12/2/2011 17:59   
524369    580546     23406            CHECK       -21   12/5/2011 9:27   
524370    580547     21201              ???      -390   12/5/2011 9:29   
524472    580561     22043           dotcom        -9  12/5/2011 10:25   
532723    580990     21109              wet       -48  12/6/2011 16:54   
532724    580991     22481        wet boxes      -288  12/6/2011 16:55   
535134    581190    84801A            check      -102  12/7/2011 16:55   
535321    581198     22025            check       -30  12/7/2011 18:26   
535322    581199     84581              NaN        -2  12/7/2011 18:26   
535323    581200    84508C            check       -21  12/7/2011 18:27   
535324    581201     22217         damages?      -155  12/7/2011 18:30   
535327    581204     85104  ????damages????      -355  12/7/2011 18:32   
535328    581205     20893          damages       -55  12/7/2011 18:32   
535329    5

以上猜想被证实错误，因为还存在`Quantity`变量为负数且`InvoiceNo`不以C开头的观察值。

但根据以上输出结果，这些筛选出的观察值的`UnitPrice`观察值均为0，因此增加`UnitPrice`的条件进行验证。

In [13]:
original_data[(original_data["Quantity"] < 0)&(original_data["InvoiceNo"].str[0] != "C")&(original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当`Quantity`变量为负数的时候，观察值满足以下条件之一：
- `InvoiceNo`以C开头，表示订单被取消
- `UnitPrice`为0，说明单价为0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将`Quantity`为负数的观察值删除。

接下来筛选出`UnitPrice`为负数的观察值。

In [14]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负数的观察值都是坏账调账，不属于实际商品交易数据，因此也在数据清理步骤中也将其删除。

## 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：
- 把`InvoiceDate`变量的数据类型转换为为日期时间
- 把`CustomerID`变量的数据类型转换为字符串
- 把`Description`变量缺失的观察值删除
- 把`Country`变量值`"USA"`替换为`"United States"`
- 把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`
- 把`Quantity`变量值为负数的观察值删除
- 把`UnitPrice`变量值为负数的观察值删除

为了区分开经过清理的数据和原始的数据，我们创建新的变量`cleaned_data`，让它为`original_data`复制出的副本。我们之后的清理步骤都将被运用在`cleaned_data`上。

In [15]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

把`InvoiceDate`变量的数据类型转换为为日期时间类型：

In [ ]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

把`CustomerID`变量的数据类型转换为字符串：

In [ ]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

把`CustomerID`变量值结尾的`".0"`删除：

In [ ]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

把`Description`变量缺失的观察值删除，并查看删除后该列空缺值个数和：

In [ ]:
cleaned_data.dropna(subset=["Description"], inplace=True)

In [ ]:
cleaned_data["Description"].isnull().sum()

把`Country`变量值`"USA"`替换为`"United States"`，并检查替换后`"USA"`变量值个数：

In [ ]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [ ]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

把`Country`变量值`"UK"`、`"U.K."`替换为`"United Kingdom"`，并检查替换后`"UK"`和`"U.K."`变量值个数：

In [ ]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})


In [ ]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

把`Quantity`变量值为负数的观察值删除，并检查替换后`Quantity`变量值为负数的个数：

In [ ]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [ ]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

把`UnitPrice`变量值为负数的观察值删除，并检查替换后`UnitPrice`变量值为负数的个数：

In [ ]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [ ]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为`e_commerce_cleaned.csv`。

In [ ]:
cleaned_data.head()

In [ ]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)

In [ ]:
pd.read_csv("e_commerce_cleaned.csv").head()